rpcq/_server.py

```diff
@@ -74,17 +83,53 @@ def exit_handler(self, f: Callable):
         """
         self._connect(endpoint)
 
-        while True:
-            try:
-                # empty_frame may either be:
-                # 1. a single null frame if the client is a REQ socket
-                # 2. an empty list (ie. no frames) if the client is a DEALER socket
-                identity, *empty_frame, msg = await self._socket.recv_multipart()
-                request = from_msgpack(msg)
-
-                asyncio.ensure_future(self._process_request(identity, empty_frame, request))
-            except Exception:
-                _log.exception('Exception thrown in Server run loop')
+        async def run_async_loop(task_list, listen_task):
+            dones, pendings = await asyncio.wait(task_list, return_when=asyncio.FIRST_COMPLETED)
+
+            # grab one "done" task to handle
+            task_list, done_list = list(pendings), list(dones)
+            done = done_list.pop()
+            task_list += done_list
+
+            if done == listen_task:
+                try:
+                    # empty_frame may either be:
+                    # 1. a single null frame if the client is a REQ socket
+                    # 2. an empty list (ie. no frames) if the client is a DEALER socket
+                    identity, *empty_frame, msg = done.result()
+                    request = from_msgpack(msg)
+
+                    # spawn a processing task
+                    task_list.append(asyncio.ensure_future(
+                        self._process_request(identity, empty_frame, request)))
+                except Exception as e:
+                    if self.serialize_exceptions:
+                        _log.exception('Exception thrown in Server run loop during request '
+                                       'reception: {}'.format(str(e)))
+                    else:
+                        raise e
+                finally:
+                    # spawn a new listen task
+                    listen_task = asyncio.ensure_future(self._socket.recv_multipart())
+                    task_list.append(listen_task)
+                    await run_async_loop(task_list, listen_task)
```

Comment:
Recursion with and Python don't play nicely. Aren't we going to blow out the recursion limit after 1000 requests?

Reply:
Bummer; I didn't know that. This is probably rework-able.

Reply:
Way too much work / too awkward / too non-pythonic to fix. I kept the try/except/finally that handles recv + deser errors & otherwise it's back to the while loop that it used to be.